### PR TITLE
8344609: Check ResourceMark nesting when allocating a GrowableArray on an alternative ResourceArea

### PR DESCRIPTION
**Pull request by reinrich**
With this change the GrowableArray nesting check is also performed if allocating from an `Arena` which in fact is a `ResourceArea`.

The additional checking can help find issue as [JDK-8328085](https://bugs.openjdk.org/browse/JDK-8328085).

The fix passed our nightly CI testing:
Tier 1-4 of hotspot and jdk. All of langtools and jaxp. Renaissance Suite and SAP specific tests.
Testing was done on the main platforms and also on Linux/PPC64le and AIX.

Besides ctw tests on ppc64 (see [JDK-8328085](https://bugs.openjdk.org/browse/JDK-8328085)) the check never failed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8344609](https://bugs.openjdk.org/browse/JDK-8344609): Check ResourceMark nesting when allocating a GrowableArray on an alternative ResourceArea (**Enhancement** - P4)

### Reviewers
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**) Review applies to [eba902e5](https://git.openjdk.org/jdk/pull/22269/files/eba902e5c4cd4aa66550460934ceaaeb2042d40e)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22269/head:pull/22269` \
`$ git checkout pull/22269`

Update a local copy of the PR: \
`$ git checkout pull/22269` \
`$ git pull https://git.openjdk.org/jdk.git pull/22269/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22269`

View PR using the GUI difftool: \
`$ git pr show -t 22269`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22269.diff">https://git.openjdk.org/jdk/pull/22269.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22269#issuecomment-2488784129)
</details>

### CODE REVIEW

**Comment by reinrich** [comment]
Thanks for the review Martin 👍

**Review by dholmes-ora — APPROVED** [review]
Seems reasonable.

What testing did you do to see if the assert fires on existing code?

Thanks

**Comment by reinrich** [comment]
> Seems reasonable.

Thanks for having a look at it.

> What testing did you do to see if the assert fires on existing code?

I forgot to update the pr description, sorry.
I've added the change about a week ago to our CI testing where we test head of jdk with and without changes from the team currently in review.
We do tier 1-4 of hotspot and jdk tests. All of langtools and jaxp. Renaissance Suite and SAP specific tests.
The testing is done nightly on the main platforms and also on Linux/PPC64le and AIX.

I had no positves from the assert except for the ctw tests on ppc64 also mentioned in the JBS issue.

**Comment by dholmes-ora** [comment]
I ran this through our CI as well - no issues in T1-4

File: src/hotspot/share/utilities/growableArray.cpp [hunk 1]
```diff
@@ -70,6 +74,11 @@ void GrowableArrayNestingCheck::on_resource_area_alloc() const {
   }
 }
 
+void GrowableArrayNestingCheck::on_arena_alloc(Arena* arena) const {
+  if ((arena->get_tag() == Arena::Tag::tag_ra) && _nesting != static_cast<ResourceArea*>(arena)->nesting()) {
+    fatal("allocation bug: GrowableArray is growing within nested ResourceMark");
```
[the remark this answers is not in the document]
Comment (reinrich) [reply src/hotspot/share/utilities/growableArray.cpp:79]: Should be a little more specific: [a new `_data` array is allocated](https://github.com/openjdk/jdk/blob/4b928167435bbf41dd00425c927da761751ca704/src/hotspot/share/utilities/growableArray.hpp#L526-L530) in the scope of a nested RM. Existing array elements are copied to the new `_data` array.

**Comment by reinrich** [comment]
Thanks for the reviews and the testing!
/integrate

**Pull request closed without merging**